sw/device/sca/aes_serial.c

```diff
@@ -298,6 +303,17 @@ static void aes_serial_single_encrypt(const uint8_t *plaintext,
   aes_send_ciphertext(false);
 }
 
+/**
+ * Advances PRNG to determine the order of measurements for fvsr-data TVLA
+ */
+static void aes_serial_advance_prng(void) {
+  bool lfsr_bit = prng_state & 0x00000001;
+  prng_state = prng_state >> 1;
+  if (lfsr_bit) {
+    prng_state ^= 0x80000057;
+  }
+}
+
```

Comment:
I don't like the idea to introduce yet another software PRNG into the framework. Right now, we have:
1. Mersenne-Twister implementation in `lib/prng.c/h`, we probably want to remove that once everything is DTR compliant. So better not use this further...
2. An LFSR in `lib/sca.c/h`.  This is mostly used for software key masking.
3. The software AES implementation in `lib/aes.c/h` we need for DTR compliance.

This here adds a 1:1 copy of 2. but without the additional support functions we have for the LFSR in `lib/sca.c/h`. If we really need a new/separate PRNG, I strongly suggest adding a second context to the LFSR in  `lib/sca.c/h` to avoid code duplication.

Also, AFAIU for fvsr-key we use a bit of the plaintext to determine the order of measurements. As the plaintexts are fully deterministic for DTR, we end up with a fully deterministic order of measurements. I assume you wanted to avoid this here and thus introduced this new PRNG. However, since the PRNG is cannot be reseeded from the host, things are still fully deterministic.

Can you please outline your thoughts a bit here? It would help me understanding what the purpose of this is.

Reply:
The issue is not that the order is deterministic (this is indeed true for other datasets as well). 
The problem is that the resulting dataset is not balanced. In other data sets we generate a random bit from plaintext. Since fixed-plaintext is constant in this case this would always result in bit 1, i.e. there will always be a random sample after a fixed sample.
Prob(fixed after random) = 50 %
Prob(random after fixed) = 100 %
Therefore: Prob(fixed) = 1/3 and Prob(random) = 2/3

In order to end up with a balanced data set, we need a way to pick fixed and random samples in a random way.
I've now made changes to re-use existing LFSR functions and I've factored these changes into a separate commit.

Reply:
Thanks for explaining! This makes a lot of sense. And the way to implemented the second LFSR context is elegant. Nice work!